### PR TITLE
use cache dir as a default remote when importing from local repo

### LINKED ISSUE #2599: `import`: get from source project cache if `url/path` are in the local system
UPDATE: Jump to https://github.com/iterative/dvc/issues/2599#issuecomment-566346857 (notice also https://github.com/iterative/dvc/issues/2599#issuecomment-700337141)

```
DVC version: 0.62.1
Python version: 3.7.3
Platform: Darwin-18.7.0-x86_64-i386-64bit
Binary: False
Filesystem type (workspace): ('apfs', '/dev/disk1s1')
```

I create a DVC project in `/path/to/data-reg` and `dvc add` a `data/` directory in it.

Then I create another DVC project in `/path/to/project` in the same file system and try to import `data/` but get:

```console
$ dvc import /path/to/data-reg data
Importing 'data (../data-reg-test)' -> 'data'
Missing cache for directory '../../../../private/var/folders/_c/3mt_xn_d4xl2ddsx2m98h_r40000gn/T/tmpkq6tyek3dvc-repo/data'.
Cache for files inside will be lost. Would you like to continue? Use '-f' to force. [y/n] y
ERROR: failed to import 'data' from '/path/to/data-reg'. - config file error: no remote specified. Setup default remote with
    dvc config core.remote <name>
or use:
    dvc pull -r <name>


Having any troubles? Hit us up at https://dvc.org/support, we are always happy to help!
```

A few problems:

- ~~The first part of the message is completely mysterious to me. I don't understand what cache dir is missing and obviously that path isn't friendly.~~
- ~~I also don't know what it's asking me to confirm so I just say `y` and move on 😋~~
- The "no remote specified" is also kind of cryptic (doesn't specify which project is missing a remote) but OK, I get it: I didn't push `data/` to a remote in the first project, so the new project can't download it. However (**main point of this issue**) it could simply copy it from the source project's cache, since they're on the same file system. This would be a nice-to-have feature I think. In fact it should probably try this before the default remote, for performance.
- ~~There's 2 new lines after `dvc pull -r <name>` in the error output.~~

UPDATE: Moved all the UI-related stuff to #2602 (per Ivan's https://github.com/iterative/dvc/issues/2599#issuecomment-541292359 below).

### PR DESCRIPTION
**Pull request by maykulkarni**
If the URL is local dvc repo, the import/get should fetch from the source project cache.

Fixes #2599

* [x] ❗ Have you followed the guidelines in the [Contributing to DVC](https://dvc.org/doc/user-guide/contributing/core) list?

* [ ] 📖 Check this box if this PR **does not** require [documentation](https://dvc.org/doc) updates, or if it does **and** you have created a separate PR in [dvc.org](https://github.com/iterative/dvc.org) with such updates (or at least opened an issue about it in that repo). Please link below to your PR (or issue) in the [dvc.org](https://github.com/iterative/dvc.org) repo.

* [ ] ❌ Have you checked DeepSource, CodeClimate, and other sanity checks below? We consider their findings recommendatory and don't expect everything to be addressed. Please review them carefully and fix those that actually improve code or fix bugs.

Thank you for the contribution - we'll try to review it as soon as possible. 🙏

### CODE REVIEW

File: dvc/external_repo.py [hunk 1]
```diff
@@ -62,16 +63,29 @@ def _external_repo(url=None, rev=None, cache_dir=None):
...
     # Adjust new clone/copy to fit rev and cache_dir
     repo = Repo(new_path)
+    # Adjust original repo for pointing remote towards its' cache
+    original_repo = Repo(url)
+    rconfig = RemoteConfig(original_repo.config)
     try:
         if rev is not None:
             repo.scm.checkout(rev)
 
+        if not _is_local(url) and not _remote_config_exists(rconfig):
```
Comment (efiop) [inline dvc/external_repo.py]: You can simply use `os.path.isdir(url)` instead of `_is_local`.

File: dvc/external_repo.py [hunk 2]
```diff
@@ -62,16 +63,29 @@ def _external_repo(url=None, rev=None, cache_dir=None):
 
     # Adjust new clone/copy to fit rev and cache_dir
     repo = Repo(new_path)
+    # Adjust original repo for pointing remote towards its' cache
+    original_repo = Repo(url)
```
Comment (efiop) [inline dvc/external_repo.py]: This won't work if url is not a local existing directory. It should really be under `if os.path.isdir(url)..` down below.

File: dvc/external_repo.py [hunk 3]
```diff
@@ -62,16 +63,29 @@ def _external_repo(url=None, rev=None, cache_dir=None):
...
+    rconfig = RemoteConfig(original_repo.config)
```
Comment (efiop) [inline dvc/external_repo.py]: Same here.

File: tests/func/test_external_repo.py [hunk 4]
```diff
@@ -26,3 +27,11 @@ def test_external_repo(erepo):
             assert path_isin(repo.cache.local.cache_dir, repo.root_dir)
 
         assert mock.call_count == 1
+
+
+def test_external_repo_import_without_remote(erepo, dvc_repo):
```
Comment (efiop) [inline tests/func/test_external_repo.py]: But `erepo` has default remote setup, so your test doesn't work at all. You need to remove default remote setting logic from erepo fixture, as it won't be needed, since you've introduced it in _external_repo in this PR.

File: dvc/external_repo.py [hunk 5]
```diff
@@ -62,16 +62,33 @@ def _external_repo(url=None, rev=None, cache_dir=None):
 
     # Adjust new clone/copy to fit rev and cache_dir
     repo = Repo(new_path)
+    original_repo = None
     try:
         if rev is not None:
             repo.scm.checkout(rev)
+        # check if the URL is local and no default remote
+        # add default remote pointing to the original repo's cache location
+        if not os.path.isdir(url):
```
Comment (efiop) [inline dvc/external_repo.py]: maybe
```suggestion
        if os.path.isdir(url):
```
?

File: dvc/external_repo.py [hunk 6]
```diff
@@ -62,16 +62,33 @@ def _external_repo(url=None, rev=None, cache_dir=None):
 
     # Adjust new clone/copy to fit rev and cache_dir
     repo = Repo(new_path)
+    original_repo = None
```
Comment (efiop) [inline dvc/external_repo.py]: No need to spread the context of that so widely, better move original_repo closing logic under the `if os.path.isdir()`

File: dvc/external_repo.py [hunk 7]
```diff
@@ -62,16 +62,33 @@ def _external_repo(url=None, rev=None, cache_dir=None):
...
+        # add default remote pointing to the original repo's cache location
+        if not os.path.isdir(url):
+            # Adjust original repo for pointing remote towards its' cache
+            original_repo = Repo(url)
+            rconfig = RemoteConfig(original_repo.config)
+            if not _remote_config_exists(rconfig):
+                rconfig.add(
+                    "upstream",
+                    original_repo.cache.local.cache_dir,
+                    default=True,
```
Comment (efiop) [inline dvc/external_repo.py]: Better set it at the local level, same as we do with `cache_dir` down below.
```suggestion
                    default=True,
                    level=Config.LEVEL_LOCAL,
```

File: dvc/external_repo.py [hunk 8]
```diff
@@ -62,16 +62,33 @@ def _external_repo(url=None, rev=None, cache_dir=None):
...
         if rev is not None:
             repo.scm.checkout(rev)
+        # check if the URL is local and no default remote
+        # add default remote pointing to the original repo's cache location
+        if not os.path.isdir(url):
+            # Adjust original repo for pointing remote towards its' cache
+            original_repo = Repo(url)
+            rconfig = RemoteConfig(original_repo.config)
+            if not _remote_config_exists(rconfig):
+                rconfig.add(
```
Comment (efiop) [inline dvc/external_repo.py]: You shouldn't modify original repo, you should modify our clone, same as we do with `cache_dir` down below.

File: dvc/external_repo.py [hunk 9]
```diff
@@ -62,16 +62,33 @@ def _external_repo(url=None, rev=None, cache_dir=None):
...
+    original_repo = None
     try:
         if rev is not None:
             repo.scm.checkout(rev)
+        # check if the URL is local and no default remote
+        # add default remote pointing to the original repo's cache location
+        if not os.path.isdir(url):
+            # Adjust original repo for pointing remote towards its' cache
+            original_repo = Repo(url)
+            rconfig = RemoteConfig(original_repo.config)
```
Comment (efiop) [inline dvc/external_repo.py]: No need to read original repo directly, we have our clone already (`repo`) and we can use it to check if default remote is set and can then set default remote in it, we def shouldn't modify the original repo directly.

File: dvc/external_repo.py [hunk 10]
```diff
@@ -62,16 +62,33 @@ def _external_repo(url=None, rev=None, cache_dir=None):
...
+            if not _remote_config_exists(rconfig):
+                rconfig.add(
+                    "upstream",
+                    original_repo.cache.local.cache_dir,
+                    default=True,
+                )
+                original_repo.scm.add([original_repo.config.config_file])
+                original_repo.scm.commit("add remote")
```
Comment (efiop) [inline dvc/external_repo.py]: This doesn't work. You already have a clone of the original repo, and now you are modifying the original repo, while not updating the clone. You need to adjust the clone and you can do it without committing, see comment above and see how we do that with `cache_dir` down below.

File: tests/func/test_external_repo.py [hunk 11]
```diff
@@ -26,3 +27,11 @@ def test_external_repo(erepo):
...
+    src = erepo.CODE
+    dst = dvc_repo.root_dir
+
+    Repo.get(erepo.root_dir, src, dst)
+    assert os.path.exists(dst + "/" + erepo.CODE)
```
Comment (efiop) [inline tests/func/test_external_repo.py]:
```suggestion
    assert os.path.exists(os.path.join(dst, erepo.CODE))
```

File: dvc/external_repo.py [hunk 12]
```diff
@@ -69,6 +69,18 @@ def _external_repo(url=None, rev=None, cache_dir=None):
...
     repo = Repo(new_path)
     try:
+        # check if the URL is local and no default remote is present
+        # add default remote pointing to the original repo's cache location
+        if os.path.isdir(url):
+            rconfig = RemoteConfig(repo.config)
+            if not _remote_config_exists(rconfig):
+                rconfig.add(
+                    "upstream",
+                    repo.cache.local.cache_dir,
```
Comment (efiop) [inline dvc/external_repo.py]: You should point to the _original repo_ cache location, not to the cache of the cloned repo, since it is empty.

File: dvc/external_repo.py [hunk 13]
```diff
@@ -113,3 +127,19 @@ def _clone_repo(url, path):
 
     git = Git.clone(url, path)
     git.close()
+
+
+def _remote_config_exists(rconfig):
```
Comment (pared) [inline dvc/external_repo.py]: I think this method could be a part of RemoteConfig

File: dvc/external_repo.py (hunk 8 above)
Comment (pared) [reply dvc/external_repo.py]: This is not resolved yet, please take a note the you are modyfing `original_repo` config, while you should modify temporary one.

Comment (maykulkarni) [reply dvc/external_repo.py]: @pared `rconfig` is the remote config of the temporary repo and not the original one.

Comment (maykulkarni) [reply dvc/external_repo.py]: @pared I think you're commenting on an older version it is fixed in the latest version. Sorry I didn't mark this conversation as resolved.

File: tests/conftest.py [hunk 14]
```diff
@@ -148,11 +147,6 @@ def erepo(repo_dir):
         repo.dvc.scm.add([stage_foo.path, stage_bar.path, stage_data_dir.path])
         repo.dvc.scm.commit("init repo")
 
-        rconfig = RemoteConfig(repo.dvc.config)
-        rconfig.add("upstream", repo.dvc.cache.local.cache_dir, default=True)
-        repo.dvc.scm.add([repo.dvc.config.config_file])
-        repo.dvc.scm.commit("add remote")
-
```
Comment (pared) [inline tests/conftest.py]: +1, nice

File: tests/func/test_get.py [hunk 15]
```diff
@@ -101,7 +101,6 @@ def test_get_from_non_dvc_master(erepo, tmp_path, monkeypatch, caplog):
     with caplog.at_level(logging.INFO, logger="dvc"):
         Repo.get(erepo._root_dir, erepo.FOO, out=imported_file, rev="branch")
 
-    assert caplog.text == ""
```
Comment (pared) [inline tests/func/test_get.py]: Ok, so this definitely cannot be left like that:
1. if you are not checking caplog text, I would just remove caplog from this test.
2. In my opinion we have a problem here, that is probably not related to this issue:
- now every time someone will try to import from local with no cache, there will be information logged
`Setting 'upstream' to default`, which will definitely be confusing, as user did not set any remote during operation, and also, this operation is performed for temporary repo, and that will be even more confusing, as nothing in original repo will change, while information will be printed. @dmpetrov, do you agree with me?
What I think we should do is comment it out with `FIXME` comment above, and create issue on supressing at least some messages from `external_repo`. @efiop?

File: dvc/remote/config.py [hunk 16]
```diff
@@ -156,3 +156,18 @@ def get_default(self, level=None):
         return self.config.get(
             Config.SECTION_CORE, Config.SECTION_CORE_REMOTE, level=level
         )
+
+    def remote_config_exists(self):
```
[the remark this answers is not in the document]
Comment (maykulkarni) [reply dvc/remote/config.py]: @pared how about `default_remote_exists`?

File: dvc/external_repo.py [hunk 17]
```diff
@@ -69,6 +69,20 @@ def _external_repo(url=None, rev=None, cache_dir=None):
 
     repo = Repo(new_path)
     try:
+        # check if the URL is local and no default remote is present
+        # add default remote pointing to the original repo's cache location
+        if os.path.isdir(url):
```
[the remark this answers is not in the document]
Comment (maykulkarni) [reply dvc/external_repo.py]: @pared that works perfectly with Linux, but on windows `urlparse("C://my//path")` gives me schema as `C`.

[the remark this answers is not in the document]
Comment (efiop) [reply dvc/external_repo.py]: @pared `isdir` is enough here, as we only need to check if the dir exists with that url, no need to bother with schemas.

File: tests/func/test_get.py (hunk 15 above)
Comment (efiop) [reply tests/func/test_get.py]: @pared Good catch! For now, we could move that logging message from `RemoteConfig.add` to `dvc/command/remote.py:RemoteAdd` so it only shows up during CLI, as it was accidentaly moved from there during the last refactoring. Don't see any other place where we need to show that message, so seems like a good solution to me. What do you think?

File: dvc/external_repo.py [hunk 18]
```diff
@@ -69,6 +69,20 @@ def _external_repo(url=None, rev=None, cache_dir=None):
...
+            rconfig = RemoteConfig(repo.config)
+            if not rconfig.remote_config_exists():
+                original_repo = Repo(url)
+                rconfig.add(
```
Comment (efiop) [inline dvc/external_repo.py]: Should be
```
                try:
                    rconfig.add(...)
                finally:
                    original_repo.close()
```

File: dvc/external_repo.py [hunk 19]
```diff
@@ -69,6 +69,20 @@ def _external_repo(url=None, rev=None, cache_dir=None):
...
+                    "upstream",
```
Comment (efiop) [inline dvc/external_repo.py]: Would call it something like
```suggestion
                    "auto-generated",
```
or something else. `upstream` is too generic, it might conflict with existing sections named like that, so it is better to reflect "auto-generated" nature of this setting. Maybe there are better names though.

File: tests/func/test_external_repo.py [hunk 20]
```diff
@@ -26,3 +28,13 @@ def test_external_repo(erepo):
             assert path_isin(repo.cache.local.cache_dir, repo.root_dir)
 
         assert mock.call_count == 1
+
+
+def test_external_repo_import_without_remote(erepo, dvc_repo):
```
Comment (efiop) [inline tests/func/test_external_repo.py]: Btw, seems like we don't even need this test, as it is automatically tested by the erepo fixture used in all other tests.

File: dvc/remote/config.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (pared) [reply dvc/remote/config.py]: @maykulkarni, sounds good!

File: tests/func/test_get.py (hunk 15 above)
Comment (pared) [reply tests/func/test_get.py]: @efiop sounds good to me, only other way that I see would be to somehow supress logging in external repo. And just by saying it out loud I feel the cringe :D

Comment (pared) [reply tests/func/test_get.py]: @maykulkarni I guess the consensus here is that this check should stay, we need to move the logging from `dvc/remote/config.py` to `CmdRemoteAdd`.

File: dvc/remote/config.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (efiop) [reply dvc/remote/config.py]: How about `default_remote_set`? Since we don't really check if it actually exists, but rather that it is set. 🙂

File: dvc/remote/config.py [hunk 21]
```diff
@@ -156,3 +155,18 @@ def get_default(self, level=None):
         return self.config.get(
             Config.SECTION_CORE, Config.SECTION_CORE_REMOTE, level=level
         )
+
+    def remote_remote_set(self):
```
Comment (pared) [inline dvc/remote/config.py]: remote_remote, typo? :)

File: dvc/remote/config.py [hunk 22]
```diff
@@ -156,3 +155,18 @@ def get_default(self, level=None):
...
+        """
+        Checks if default remote config is present.
+        Args:
+            rconfig: a remote config
+
+        Returns:
+            True if the remote config exists, else False
```
Comment (pared) [inline dvc/remote/config.py]: True if default remote config is set, else False?

File: dvc/remote/config.py (hunk 21 above)
Comment (maykulkarni) [reply dvc/remote/config.py]: Geez sorry my bad

File: dvc/external_repo.py [hunk 23]
```diff
@@ -69,6 +69,22 @@ def _external_repo(url=None, rev=None, cache_dir=None):
...
     try:
+        # check if the URL is local and no default remote is present
+        # add default remote pointing to the original repo's cache location
+        if os.path.isdir(url):
+            rconfig = RemoteConfig(repo.config)
+            if not rconfig.default_remote_set():
+                original_repo = Repo(url)
+                try:
+                    rconfig.add(
+                        "aut-generated-upstream",
```
Comment (efiop) [inline dvc/external_repo.py]:
```suggestion
                        "auto-generated-upstream",
```

File: dvc/remote/config.py [hunk 24]
```diff
@@ -156,3 +155,18 @@ def get_default(self, level=None):
...
+            rconfig: a remote config
+
+        Returns:
+            True if the default remote config is set, else False
+        """
+        try:
+            self.get_default()
+            return True
+        except ConfigError:
+            return False
```
Comment (efiop) [inline dvc/remote/config.py]: Let's not have it here, as no one is using it. Let's keep this helper in dvc/external_repo.py

File: dvc/external_repo.py [hunk 25]
```diff
@@ -4,11 +4,11 @@
 import tempfile
 from contextlib import contextmanager
 from distutils.dir_util import copy_tree
-
```
Comment (efiop) [inline dvc/external_repo.py]: Please don't remove this, it separates std from 3rd party imports. And, well, touching unrelated lines is no bueno in general 🙂

File: dvc/external_repo.py [hunk 26]
```diff
@@ -4,11 +4,11 @@
...
 from funcy import retry
 
-from dvc.config import NoRemoteError
```
Comment (efiop) [inline dvc/external_repo.py]: Same here, no need to move it really. It breaks the git blame.

Comment (maykulkarni) [reply dvc/external_repo.py]: @efiop codeclimate asked me to do it :-X

File: dvc/external_repo.py (hunk 25 above)
Comment (maykulkarni) [reply dvc/external_repo.py]: @efiop deepsource asked me to do it :-X

File: dvc/external_repo.py (hunk 26 above)
Comment (efiop) [reply dvc/external_repo.py]: @maykulkarni It is purely advisory, no need to make it fully pass.

File: dvc/external_repo.py (hunk 25 above)
Comment (efiop) [reply dvc/external_repo.py]: @maykulkarni It is purely advisory, no need to make it fully pass.

**Comment by efiop** [comment]
@maykulkarni Could you also create a ticket for the docs or create a PR for docs to reflect this new logic?

File: dvc/external_repo.py [hunk 27]
```diff
@@ -69,6 +70,22 @@ def _external_repo(url=None, rev=None, cache_dir=None):
 
     repo = Repo(new_path)
     try:
+        # check if the URL is local and no default remote is present
+        # add default remote pointing to the original repo's cache location
+        if os.path.isdir(url):
+            rconfig = RemoteConfig(repo.config)
+            if not _default_remote_set(rconfig):
```
Comment (jorgeorpinel) [inline dvc/external_repo.py]: @maykulkarni @efiop does this logic mean that if there's a default remote setup, still DVC skips the local cache? Because I think it should try the local cache first regardless of remotes.

Comment (efiop) [reply dvc/external_repo.py]: @jorgeorpinel Yes, precisely. It is hard for us to try both right now, as we don't have that kind of cascading remote logic implemented. I would wait for someone to ask for this.

Comment (jorgeorpinel) [reply dvc/external_repo.py]: Sorry, I'm not sure I understand:

1. Every DVC repo has a cache directory right? So in theory it can always be the first place to get stuff from when using get/import from another local project. Why does this require complicated remote checks?

2. Does this PR really close #2599 per that issue's description and discussion?

Comment (efiop) [reply dvc/external_repo.py]: 1. Yes, but "first place to get stuff from" means that there is a second place, and our code is not ready to cascade from first to second yet, if it can't find all the stuff it needs in the local cache dir.

2. From what we've understood initially - yes, but now it looks like we didn't quite understand each other.

Comment (jorgeorpinel) [reply dvc/external_repo.py]: I see. OK!
